src/main/java/org/broadinstitute/hellbender/engine/datasources/ReadsHeaderSource.java

```diff
@@ -0,0 +1,75 @@
+package org.broadinstitute.hellbender.engine.datasources;
+
+import com.google.api.services.storage.Storage;
+import com.google.cloud.genomics.dataflow.readers.bam.BAMIO;
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SamReader;
+import htsjdk.samtools.SamReaderFactory;
+import htsjdk.samtools.ValidationStringency;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.broadinstitute.hellbender.engine.spark.AuthHolder;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.Utils;
+import org.broadinstitute.hellbender.utils.gcs.BucketUtils;
+import org.seqdoop.hadoop_bam.util.SAMHeaderReader;
+
+import java.io.File;
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+
+/**
+ * Class to read BAM headers. May be expanded to include BAM contents later, if useful.
+ */
+public final class ReadsHeaderSource {
+    private final String bam;
+    private final boolean cloudStorageUrl;
+    private final boolean hadoopUrl;
+    private AuthHolder auth;
+
+
+    /**
+     * When using this constructor, none of the PCollection-generating code will be available.
+     *
+     * @param bam A file path or a google bucket identifier to a bam file to read
+     * @param auth - the authentication for this job.
+     */
+    public ReadsHeaderSource(String bam, AuthHolder auth) {
+        this.bam = Utils.nonNull(bam);
+
+        cloudStorageUrl = BucketUtils.isCloudStorageUrl(bam);
+        if (cloudStorageUrl) {
+            Utils.nonNull(auth);
+        }
+        hadoopUrl = BucketUtils.isHadoopUrl(bam);
+        if (cloudStorageUrl) {
+            this.auth = auth;
+        }
+    }
+
+    /**
+     * Get a SAMFileHeader to use with Reads produced by this ReadsSource
+     */
+    public SAMFileHeader getHeader() {
+        if(cloudStorageUrl) {
+            try {
+                Storage.Objects storageClient = auth.makeStorageClient();
+                final SamReader reader = BAMIO.openBAM(storageClient, bam, ValidationStringency.DEFAULT_STRINGENCY);
+                return reader.getFileHeader();
+            } catch (IOException e) {
+                throw new GATKException("Failed to read bams header from " + bam + ".", e);
+            } catch (ClassNotFoundException | GeneralSecurityException e) {
+                throw new GATKException("Failed to read bams header from " + bam + " (authentication issue?).", e);
+            }
+        } else if (hadoopUrl) {
+            try {
+                return SAMHeaderReader.readSAMHeaderFrom(new Path(bam), new Configuration());
+            } catch (IOException e) {
+                throw new GATKException("Failed to read bams header from " + bam + ".", e);
+            }
+        } else {
+            return SamReaderFactory.makeDefault().getFileHeader(new File(bam));
```

Comment:
`SamReaderFactor.makeDefault` returns a `SamReaderFactory`, counter-intuitively. There is nothing for me to close.